python.rst

```diff
@@ -2,36 +2,96 @@
  Python Style Guide
 ====================
 
-Goal
+Background
 ====
 
-The goal of this style-guide is to standardize the motif of Python at
-Oscar in a way that reduces the risk of bugs while increasing our
-ability to utilize the richer features of Python when necessary. In
-addition, it reduces the risk exposure from the many third party
-libraries we use. Finally, if it is true that reading code is 90% of
-programming, then we should favor readability over writability.
+The goal of this document is to guide the style of Python code at
+Oscar to reduce bugs and improve readability.
 
-Example: `Import packages and modules only, do not import objects or
-functions directly`_ may seem constrictive or verbose, but it allows
-objects in module namespaces to be mutated during execution without
-unexpected behavior due to object duplication. Examples of this in our
-current code-base are sqlalchemy-continuum and tagging.
+Many teams use the `black <https://github.com/psf/black>`_
+auto-formatter to avoid having to worry (or even worse, argue!) about
+formatting. We recommend using it, but it does not cover everything
+here.
 
-Code of Conduct
+Code of conduct
 ===============
 
 - Be polite and respectful.
 - Don’t make assumptions—be clear and concise.
-- Your co workers deserve a response to comments, questions or
-  concerns, even if it is to respectfully decline or dissent.
+- Your coworkers deserve a response to comments, questions or concerns, even if it is to respectfully decline or dissent.
 - Boy Scout Rule: Always check a module in cleaner than you checked it
   out. Clean-up and compliance to the style guide is the
-  responsibility of whomever touches the code.
+  responsibility of whoever touches the code.
+- Don't Boil the Ocean: if you weren't already working on a module for a diff, you
+  probably shouldn't be cleaning it up. It confuses reviewers and makes it harder
+  to spot bugs. (Feel free to open a separate diff for the clean-up, though!)
+- Don't be afraid to ask for help or clarification. We're all
+  learning.
 
 Language Rules
 ==============
 
+Testing
+-------
+
+Use pytest-style asserts in unit tests
```

Comment:
Oh, also, if this is something we are going to do, I really think we need to help people out with a code mode. Best I can tell this requires updating over 150k lines of code.

Reply:
pytest-style asserts make tests easier to write, easier to read, and produce better error messages on failure.

Tests will be easier to write because, instead of needing to remember multiple assertion method names, devs can write them like a normal python if-statement. Even devs who know the methods regularly misuse them—for example, the unittest docs note that it's usually not necessary to invoke type-specific equality funcs (assertSetEqual, etc) directly because assertEqual already calls them, but our python codebase uses them ~1800 times.

The "looks just like python" also makes it easier to read. As a bonus, it's more concise, letting the actual check take up more of the line instead of being stuck in parens.

Less API misuse means better error messages. For example, our codebase has >1000 usages of `self.assertTrue(x in y)` that should be written in unittest as `self.assertIn(x, y)`—if devs just wrote `assert x in y`, they would have gotten an error message with proper introspection instead of just the boolean failure.

The pytest docs tout the theoretical benefits of writing comparisons as idiomatic python, and that argument lines up with the evidence of unittest API misuse in our codebase.

Reply:
A codemod would be nice, but I also don't expect people to do this wholesale to their codebases. On Sky, we write new tests like this + update old tests whenever we're making other edits to them. Are you imagining a codemod for that? a big bang? something else?

Reply:
One nice thing about making mass changes to unit tests is that you know immediately and without much overhead whether or not there's an issue. So, I wouldn't be opposed to a one-shot deal, though, I suspect you're correct that teams may not want to adopt this all at once.

By writing a codemod using something like lib2to3, we have the ability to let people run the codemod for single files, whole directories, or the whole codebase. Here's an example: https://phabricator.hioscar.com/D218510

People can then choose whatever rollout strategy they like.

Reply:
pytest does support unittest; oscar already runs all its tests via pytest.

In particular, devs can use unittest's `setup`, `mock`, and so on without having to use unittest's assert methods—which is what I'm expecting/recommending here.

Even if one made the case for switching to pytest-native versions of those features, it'd be much less clarity-improvement-bang for much more having-to-learn-a-new-API-buck than switching asserts is.

Reply:
Narrowed the scope of the parameterized suggestion + recommended prioritizing consistency; closing